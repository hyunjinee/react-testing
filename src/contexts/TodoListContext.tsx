import { createContext, useEffect, useState } from 'react';

interface Context {
  readonly todoList: string[];
  readonly addTodo: (todo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

export const TodoListContext = createContext<Context>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface Props {
  children: JSX.Element | JSX.Element[];
}

export const TodoListProvider: React.FC<Props> = ({ children }) => {
  const [todoList, setTodoList] = useState<string[]>([]);

  useEffect(() => {
    const list = localStorage.getItem('todoList');
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  const addTodo = (todo: string) => {
    if (todo) {
      const newList = [...todoList, todo];
      localStorage.setItem('todoList', JSON.stringify(newList));
      setTodoList(newList);
    }
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem('todoList', JSON.stringify(list));
    setTodoList(list);
  };

  return (
    <TodoListContext.Provider
      value={{
        todoList,
        addTodo,
        deleteTodo,
      }}
    >
      {children}
    </TodoListContext.Provider>
  );
};
