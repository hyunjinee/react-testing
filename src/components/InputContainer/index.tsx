import { useContext, useState } from 'react';

import Button from 'components/common/Button';
import Input from 'components/common/Input';
import { TodoListContext } from 'contexts/TodoListContext';
import { Container } from './styles';

interface Props {
  readonly onAdd?: () => void;
}

const InputContainer: React.FC<Props> = ({ onAdd }) => {
  const [todo, setTodo] = useState('');
  const { addTodo } = useContext(TodoListContext);

  return (
    <Container>
      <Input
        placeholder="할 일을 입력해 주세요."
        value={todo}
        onChange={setTodo}
      />
      <Button
        label="추가"
        onClick={() => {
          addTodo(todo);
          setTodo('');
          if (todo && typeof onAdd === 'function') {
            onAdd();
          }
        }}
      />
    </Container>
  );
};

export default InputContainer;
