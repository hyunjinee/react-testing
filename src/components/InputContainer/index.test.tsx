import { fireEvent, render, screen } from '@testing-library/react';
import { TodoListProvider } from 'contexts/TodoListContext';
import InputContainer from '.';

describe('<InputContainer/>', () => {
  it('renders component correctly', () => {
    const { container } = render(<InputContainer />);

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요.');
    expect(input).toBeInTheDocument();
    const button = screen.getByText('추가');
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('empties data after adding data', () => {
    render(<InputContainer />);

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요.',
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: '아이폰 사기' } });
    expect(input.value).toBe('아이폰 사기');
    fireEvent.click(button);
    expect(input.value).toBe('');
  });

  it('adds input data to localStorage via context', () => {
    render(
      <TodoListProvider>
        <InputContainer />
      </TodoListProvider>,
    );

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요.');
    const button = screen.getByText('추가');

    expect(localStorage.getItem('todoList')).toBeNull();

    fireEvent.change(input, { target: { value: '아이폰 사기' } });
    fireEvent.click(button);

    expect(localStorage.getItem('todoList')).toBe(
      JSON.stringify(['아이폰 사기']),
    );
  });

  it('class the onAdd function when the user clicks Add button', () => {
    const handleClick = jest.fn();
    render(<InputContainer onAdd={handleClick} />);

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요.');
    const button = screen.getByText('추가');
    expect(handleClick).not.toHaveBeenCalled();

    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(0);

    fireEvent.change(input, { target: { value: '아이폰 사기' } });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
