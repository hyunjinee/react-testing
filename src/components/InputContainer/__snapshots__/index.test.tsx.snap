// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer/> renders component correctly 1`] = `
<div>
  <div
    class="sc-hKMtZM ggUkCk"
  >
    <input
      class="sc-dkzDqf kpbFyQ"
      placeholder="할 일을 입력해 주세요."
      value=""
    />
    <div
      class="sc-bczRLJ zgHZe"
    >
      <div
        class="sc-gsnTZi fSQIOg"
      >
        추가
      </div>
    </div>
  </div>
</div>
`;
